lib/Sema/CSBindings.cpp

```diff
@@ -1722,56 +1724,95 @@ void BindingSet::dump(llvm::raw_ostream &out, unsigned indent) const {
 
   auto numDefaultable = getNumViableDefaultableBindings();
   if (numDefaultable > 0)
-    out << "#defaultable_bindings: " << numDefaultable << " ";
+    out << "[#defaultable_bindings: " << numDefaultable << "] ";
 
-  auto printBinding = [&](const PotentialBinding &binding) {
-    auto type = binding.BindingType;
-    switch (binding.Kind) {
-    case AllowedBindingKind::Exact:
-      break;
+  struct PrintableBinding {
+  private:
+    enum class BindingKind { Exact, Subtypes, Supertypes, Literal };
+    BindingKind Kind;
+    Type BindingType;
+    PrintableBinding(BindingKind kind, Type bindingType)
+        : Kind(kind), BindingType(bindingType) {}
 
-    case AllowedBindingKind::Subtypes:
-      out << "(subtypes of) ";
-      break;
+  public:
+    static PrintableBinding supertypesOf(Type binding) {
+      return PrintableBinding{BindingKind::Supertypes, binding};
+    }
+    
+    static PrintableBinding subtypesOf(Type binding) {
+      return PrintableBinding{BindingKind::Subtypes, binding};
+    }
+    
+    static PrintableBinding exact(Type binding) {
+      return PrintableBinding{BindingKind::Exact, binding};
+    }
+    
+    static PrintableBinding literalDefaultType(Type binding) {
+      return PrintableBinding{BindingKind::Literal, binding};
+    }
 
-    case AllowedBindingKind::Supertypes:
-      out << "(supertypes of) ";
-      break;
+    void print(llvm::raw_ostream &out, const PrintOptions &PO,
+               unsigned indent = 0) const {
+      switch (Kind) {
+      case BindingKind::Exact:
+        break;
+      case BindingKind::Subtypes:
+        out << "(subtypes of) ";
+        break;
+      case BindingKind::Supertypes:
+        out << "(supertypes of) ";
+        break;
+      case BindingKind::Literal:
+        out << "(default type of literal) ";
+        break;
+      }
+      BindingType.print(out);
     }
-    if (auto *literal = binding.getDefaultedLiteralProtocol())
-      out << "(default from " << literal->getName() << ") ";
-    out << type.getString(PO);
   };
 
   out << "[with possible bindings: ";
-  interleave(Bindings, printBinding, [&]() { out << "; "; });
-  if (!Literals.empty()) {
-    std::vector<std::string> defaultLiterals;
-    for (const auto &literal : Literals) {
-      if (literal.second.viableAsBinding()) {
-        auto defaultWithType = "(default type of literal) " +
-                               literal.second.getDefaultType().getString(PO);
-        defaultLiterals.push_back(defaultWithType);
-      }
+  SmallVector<PrintableBinding, 2> potentialBindings;
+  for (const auto &binding : Bindings) {
+    switch (binding.Kind) {
+    case AllowedBindingKind::Exact:
+      break;
```

Comment:
I think it’s possible that this set has exact bindings too i.e. something that is inferred from equality constraint, optional object, or similar…